server/api/setUserPuzzleMeta.js

```diff
@@ -0,0 +1,120 @@
+  // get the data we need; we should already be authorized so there should already be a session
+  const { userID } = req.session;
+  const { progress, puzzle_id: puzzleID, time } = req.body;
+
+  // safety net - check for required key existence
+  if (
+    !("progress" in req.body && "puzzle_id" in req.body && "time" in req.body)
+  ) {
+    res
+      .status(400)
+      .send(
+        `Error: Malformed input - required key "progress", "puzzle_id", or "time" not found in request body.`
+      );
+    return;
+  }
+
+  // safety net - check for typing
+  if (isNaN(progress) || isNaN(puzzleID) || isNaN(time)) {
+    res
+      .status(400)
+      .send(
+        `Error: Malformed input - key "progress", "puzzle_id", or "time" is not a Number.`
+      );
+    return;
+  }
+
+  // check that the specific User_Puzzle exists
+  db.get(
+    "SELECT * FROM User_Puzzle WHERE user_id = ? AND puzzle_id = ?",
+    [userID, puzzleID],
+    function (err, row) {
+      if (err) {
+        res.status(500).send(err);
+        return;
+      }
+
+      if (row) {
+        // update the value in the database
+        db.run(
+          "UPDATE User_Puzzle SET progress = ?, time = ? WHERE user_id = ? AND puzzle_id = ?",
+          [progress, time, userID, puzzleID],
+          function (err) {
+            // throw errors if there's an issue
+            if (err) {
+              res.status(500).send(err);
+              return;
+            }
+
+            // if it's done then we're fine!
+            res
+              .status(200)
+              .send(
+                `User_Puzzle relation between user_id ${userID} and puzzle_id ${puzzleID} successfully updated.`
+              );
+          }
+        );
+      } else {
+        // User_Puzzle does not exist; make User_Puzzle
+        // but first check that the Puzzle we want to insert exists
+        db.get(
+          "SELECT * FROM Puzzle WHERE puzzle_id = ?",
+          [puzzleID],
+          function (err, row) {
+            if (err) {
+              res.status(500).send(err);
+              return;
+            }
+
+            if (!row) {
+              res
+                .status(400)
+                .send(
+                  `Puzzle with "puzzle_id" ${puzzleID} does not exist. Not inserting.`
+                );
+            } else {
+              // now, since the Puzzle exists, we can do this
+              db.run(
+                "INSERT INTO User_Puzzle VALUES (?, ?, ?, ?)",
+                [userID, puzzleID, time, progress],
+                function (err) {
+                  if (err) {
+                    res.status(500).send(err);
+                    return;
+                  }
+
+                  // if it's done then we're fine!
+                  res
+                    .status(200)
+                    .send(
+                      `User_Puzzle relation between user_id ${userID} and puzzle_id ${puzzleID} successfully created.`
+                    );
```

Comment:
not a big deal tbh, but I'd say this could be one line 🤷‍♂️

Reply:
changed due to prettification; just how it is chief